File: Packages/mcp-unity/Server~/src/tools/prefabSaveTool.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import * as z from 'zod';
import { Logger } from '../utils/logger.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

const toolName = 'prefab_save';
const toolDescription = 'Saves a prefab asset directly, applies prefab instance overrides, or saves a scene object as a prefab asset.';

const paramsSchema = z.object({
  instanceId: z.number().optional().describe('The instance ID of the GameObject'),
  objectPath: z.string().optional().describe('The hierarchy path of the GameObject'),
  prefabPath: z.string().optional().describe("The prefab asset path to save to or update (e.g. 'Assets/Prefabs/MyPrefab.prefab')"),
  applyOverrides: z.boolean().default(false).describe('If true, apply prefab instance overrides instead of saving a prefab asset')
});

export function registerPrefabSaveTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);

  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: z.infer<typeof paramsSchema>) => {
      try {
        logger.info(`Executing tool: ${toolName}`, params);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

async function toolHandler(mcpUnity: McpUnity, params: z.infer<typeof paramsSchema>): Promise<CallToolResult> {
  const { instanceId, objectPath, prefabPath, applyOverrides } = params;

  if (instanceId === undefined && (!objectPath || objectPath.trim() === '') && (!prefabPath || prefabPath.trim() === '')) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Provide either 'prefabPath' or 'instanceId'/'objectPath'"
    );
  }

  if (applyOverrides && instanceId === undefined && (!objectPath || objectPath.trim() === '')) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Either 'instanceId' or 'objectPath' must be provided when 'applyOverrides' is true"
    );
  }

  const response = await mcpUnity.sendRequest({
    method: toolName,
    params: { instanceId, objectPath, prefabPath, applyOverrides }
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to save prefab'
    );
  }

  return {
    content: [{
      type: response.type || 'text',
      text: response.message || 'Successfully saved prefab'
    }],
    data: {
      prefabPath: response.prefabPath,
      guid: response.guid,
      instanceId: response.instanceId,
      objectPath: response.objectPath
    }
  };
}

File: Packages/mcp-unity/Editor/Tools/PrefabSaveTool.cs
using System;
using System.IO;
using UnityEngine;
using UnityEditor;
using Newtonsoft.Json.Linq;
using McpUnity.Unity;
using McpUnity.Utils;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for saving prefab assets or prefab instance overrides.
    /// </summary>
    public class PrefabSaveTool : McpToolBase
    {
        public PrefabSaveTool()
        {
            Name = "prefab_save";
            Description = "Saves a prefab asset directly, or applies prefab instance overrides, or saves a scene object as a prefab asset.";
        }

        public override JObject Execute(JObject parameters)
        {
            int? instanceId = parameters?["instanceId"]?.ToObject<int?>();
            string objectPath = parameters?["objectPath"]?.ToObject<string>();
            string prefabPath = parameters?["prefabPath"]?.ToObject<string>();
            bool applyOverrides = parameters?["applyOverrides"]?.ToObject<bool?>() ?? false;

            GameObject gameObject = null;
            string objectPathUsed = objectPath;

            if (instanceId.HasValue || !string.IsNullOrEmpty(objectPath))
            {
                JObject error = GameObjectToolUtils.FindGameObject(instanceId, objectPath, out gameObject, out objectPathUsed);
                if (error != null)
                {
                    return error;
                }
            }

            if (applyOverrides)
            {
                if (gameObject == null)
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        "Either 'instanceId' or 'objectPath' must be provided when 'applyOverrides' is true",
                        "validation_error"
                    );
                }

                GameObject prefabInstanceRoot = PrefabUtility.GetNearestPrefabInstanceRoot(gameObject);
                if (prefabInstanceRoot == null)
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        "The specified GameObject is not part of a prefab instance",
                        "validation_error"
                    );
                }

                PrefabUtility.ApplyPrefabInstance(prefabInstanceRoot, InteractionMode.UserAction);

                return new JObject
                {
                    ["success"] = true,
                    ["type"] = "text",
                    ["message"] = $"Applied prefab overrides for '{prefabInstanceRoot.name}'",
                    ["instanceId"] = McpObjectId.FromObject(prefabInstanceRoot),
                    ["objectPath"] = objectPathUsed,
                    ["prefabPath"] = PrefabUtility.GetPrefabAssetPathOfNearestInstanceRoot(prefabInstanceRoot)
                };
            }

            if (!string.IsNullOrEmpty(prefabPath))
            {
                prefabPath = NormalizeAssetPath(prefabPath.Trim());

                if (gameObject != null)
                {
                    if (PrefabUtility.IsPartOfPrefabAsset(gameObject))
                    {
                        try
                        {
                            PrefabUtility.SavePrefabAsset(gameObject);
                        }
                        catch (Exception ex)
                        {
                            return McpUnitySocketHandler.CreateErrorResponse(
                                $"Failed to save prefab asset at '{prefabPath}': {ex.Message}",
                                "save_error"
                            );
                        }
                    }
                    else
                    {
                        EnsureFolderHierarchy(Path.GetDirectoryName(prefabPath));

                        PrefabUtility.SaveAsPrefabAsset(gameObject, prefabPath, out bool prefabSaved);
                        if (!prefabSaved)
                        {
                            return McpUnitySocketHandler.CreateErrorResponse(
                                $"Failed to save scene object as prefab at '{prefabPath}'",
                                "save_error"
                            );
                        }
                    }
                }
                else
                {
                    GameObject prefab = AssetDatabase.LoadAssetAtPath<GameObject>(prefabPath);
                    if (prefab == null)
                    {
                        return McpUnitySocketHandler.CreateErrorResponse(
                            $"Prefab asset not found at path '{prefabPath}'",
                            "not_found_error"
                        );
                    }

                    try
                    {
                        PrefabUtility.SavePrefabAsset(prefab);
                    }
                    catch (Exception ex)
                    {
                        return McpUnitySocketHandler.CreateErrorResponse(
                            $"Failed to save prefab asset at '{prefabPath}': {ex.Message}",
                            "save_error"
                        );
                    }
                }

                string guid = AssetDatabase.AssetPathToGUID(prefabPath);

                return new JObject
                {
                    ["success"] = true,
                    ["type"] = "text",
                    ["message"] = $"Successfully saved prefab at '{prefabPath}'",
                    ["prefabPath"] = prefabPath,
                    ["guid"] = guid,
                    ["instanceId"] = gameObject != null ? new JValue(McpObjectId.FromObject(gameObject)) : null,
                    ["objectPath"] = objectPathUsed
                };
            }

            if (gameObject == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Either 'prefabPath' or ('instanceId'/'objectPath') must be provided",
                    "validation_error"
                );
            }

            if (PrefabUtility.IsPartOfPrefabAsset(gameObject))
            {
                try
                {
                    PrefabUtility.SavePrefabAsset(gameObject);
                }
                catch (Exception ex)
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        $"Failed to save prefab asset for '{gameObject.name}': {ex.Message}",
                        "save_error"
                    );
                }

                string assetPath = AssetDatabase.GetAssetPath(gameObject);
                return new JObject
                {
                    ["success"] = true,
                    ["type"] = "text",
                    ["message"] = $"Successfully saved prefab asset '{gameObject.name}'",
                    ["prefabPath"] = assetPath,
                    ["guid"] = AssetDatabase.AssetPathToGUID(assetPath),
                    ["instanceId"] = McpObjectId.FromObject(gameObject),
                    ["objectPath"] = objectPathUsed
                };
            }

            if (!string.IsNullOrEmpty(prefabPath))
            {
                // Handled above.
            }

            return McpUnitySocketHandler.CreateErrorResponse(
                "Parameter 'prefabPath' is required when saving a scene object without applying overrides",
                "validation_error"
            );
        }

        private static void EnsureFolderHierarchy(string folderPath)
        {
            folderPath = NormalizeAssetPath(folderPath);

            if (string.IsNullOrEmpty(folderPath) || AssetDatabase.IsValidFolder(folderPath))
            {
                return;
            }

            string parentPath = NormalizeAssetPath(Path.GetDirectoryName(folderPath));
            string folderName = Path.GetFileName(folderPath);

            if (!string.IsNullOrEmpty(parentPath) && !AssetDatabase.IsValidFolder(parentPath))
            {
                EnsureFolderHierarchy(parentPath);
            }

            if (!AssetDatabase.IsValidFolder(folderPath) && !string.IsNullOrEmpty(parentPath) && !string.IsNullOrEmpty(folderName))
            {
                AssetDatabase.CreateFolder(parentPath, folderName);
            }
        }

        private static string NormalizeAssetPath(string path)
        {
            return string.IsNullOrEmpty(path) ? path : path.Replace('\\', '/');
        }
    }
}
